pandas/tslib.pyx

```diff
@@ -416,12 +416,21 @@ class Timestamp(_Timestamp):
         cdef object result, value
 
         from pandas.tseries.frequencies import to_offset
-        unit = to_offset(freq).nanos
+        offset = to_offset(freq)
+        unit = offset.nanos
+        time_unit = offset.rule_code
         if self.tz is not None:
             value = self.tz_localize(None).value
         else:
             value = self.value
-        result = (unit * rounder(value / float(unit)).astype('i8'))
+        if time_unit == 'N':
```

Comment:
instead of this way I think you want to deal with ``offset.nanos < 1000 and offset.nanos % 1000 != 0``. 

This then would have an error case if ``offset.nanos >= 1000 and offset.nanos % 1000 != 0`` 

```
# ok!
In [11]: to_offset('1D').nanos % 1000
Out[11]: 0

# ok
In [12]: to_offset('10ms').nanos % 1000
Out[12]: 0

# ok
In [13]: to_offset('10us').nanos % 1000
Out[13]: 0

# ok (with the separate rounding)
In [14]: to_offset('10ns').nanos % 1000
Out[14]: 10

# NOT OK, this we would lose precision on! raise!
In [15]: to_offset('1010ns').nanos % 1000
Out[15]: 10
```